docs/extending/preprocessor.rst

```diff
@@ -0,0 +1,246 @@
+   not specify, it will work on all the data types given by the
+   ``get_valid_inputs`` function.
+
+.. attention::
+
+   Only basic types (*int*, *bool* and *str*), lists, tuples and dictionaries
+   are allowed as parameters. This is because the parameters are stored in
+   JSON format, and JSON only supports these types.
+
+.. code-block:: python
+
+    from typing import Literal
+
+    from numpy.typing import ArrayLike
+
+
+    ...
+
+    def __init__(
+        self,
+        fwhm: int | float | ArrayLike | Literal["fast"] | None,
+        on: str | list[str] | None = None,
+    ) -> None:
+        self.fwhm = fwhm
+        super().__init__(on=on)
+
+    ...
+
+.. caution::
+
+   Parameters of the Preprocessor must be stored as object attributes without
+   using ``_`` as prefix. This is because any attribute that starts with ``_``
+   will not be considered as a parameter and not stored as part of the metadata
+   of the Preprocessor.
+
+.. _extending_preprocessors_preprocess:
+
+Step 3: Preprocess the data
+---------------------------
+
+Finally, we will write the actual logic of the Preprocessor. This method will
+be called by ``junifer`` when needed, using the data provided by the
+DataGrabber, as configured by the user. The method ``preprocess`` has two
+arguments:
+
+* ``input``: A dictionary with the data to be used by the Preprocessor. This
+  will be the corresponding element in the :ref:`Data Object<data_object>`
+  already indexed. Thus, the dictionary has at least two keys: ``data`` and
+  ``path``. The first one contains the data, while the second one contains the
+  path to the data. The dictionary can also contain other keys, depending on the
+  data type.
+* ``extra_input``: The rest of the :ref:`Data Object<data_object>`. This is
+  useful if you want to use other data (e.g., ``Warp`` can be used to provide
+  the transformation matrix file for transformation to subject-native space).
+
+and it has two return values:
+
+* First is the ``input`` dictionary with necessary data modified. Usually, you
+  want to replace the ``input["data"]`` with the preprocessed data.
+* Second is a dictionary just like ``input`` or ``extra_input`` but with only
+  specific key-value pairs which you would like to pass down to the Markers.
+  For example, if your Preprocessor computes some mask with the preprocessed
+  data, you could pass it through this which would be added and available
+  in the Marker step with the same key you pass here. Usually, you would
+  want to pass ``None``.
+
+.. code-block:: python
+
+    from typing import Any
+
+    from nilearn import image as nimg
+
+
+    ...
+
+    def preprocess(
+        self,
+        input: dict[str, Any],
+        extra_input: dict[str, Any] | None = None,
+    ) -> tuple[dict[str, Any], dict[str, Any] | None]:
+         input["data"] = nimg.smooth_img(imgs=input["data"], fwhm=self.fwhm)
+         return input, None
+
+     ...
+
+Step 4: Finalise the Preprocessor
```

Comment:
I believe in multiple places I saw the american spelling of things, so this should be Finalize. Can we record in some central place i.e. something like "How to contribute to docs" that we aim to use the american spelling?

Reply:
Ok, i will point out the american way when i find them.